crates/autopilot/src/run_loop.rs

```diff
@@ -560,45 +557,45 @@ impl RunLoop {
                     );
                     None
                 }
+            });
+
+        // Winners are selected one by one, starting from the best solution,
+        // until `max_winners_per_auction` are selected. The solution is a winner
+        // if it swaps tokens that are not yet swapped by any other already
+        // selected winner.
+        let mut already_swapped_tokens = HashSet::new();
+        let mut winners = 0;
+        let solutions = solutions
+            .map(|participant| {
+                let swapped_tokens = participant
+                    .solution
+                    .orders()
+                    .iter()
+                    .flat_map(|(_, order)| vec![order.sell.token, order.buy.token])
```

Comment:
no need to allocate vectors here.
```suggestion
                    .flat_map(|(_, order)| [order.sell.token, order.buy.token])
```